aten/src/ATen/core/jit_type.h

```diff
@@ -391,6 +392,231 @@ struct CAFFE2_API DimensionedTensorType : public TensorType {
   int64_t dim_;
 };
 
+
+template <typename T>
+inline c10::optional<T> merge_primitive (const c10::optional<T>& a, const c10::optional<T>& b)
+{
+  if (a.has_value() && b.has_value() &&
+      a.value() == b.value()) {
+    return a;
+  }
+  return c10::optional<T>{};
+}
+
+
+struct CAFFE2_API VaryingShape {
+
+  VaryingShape(const IntArrayRef& ref)
+  : is_varying_size_(false)
+  , dims_(ref.begin(), ref.end())
+  {
+  }
+
+  VaryingShape(const std::vector<int64_t>& vec)
+  : is_varying_size_(false)
+  , dims_(vec.begin(), vec.end())
+  {
+  }
+
+  VaryingShape(size_t size, bool is_varying_size = false)
+  : is_varying_size_(is_varying_size)
+  , dims_(size)
+  {
+  }
+
+  bool operator ==(const VaryingShape &other) const
+  {
+    if (is_varying_size_ != other.is_varying_size_)
+    {
+      return false;
+    }
+
+    return dims_ == dims_;
+  }
+
+  c10::optional<int64_t>& operator [] (int i)
+  {
+      if (is_varying_size())
+      {
+        throw std::runtime_error("Rank isn't fixed");
+      }
+      return dims_[i];
+  }
+
+  const c10::optional<int64_t>& operator [] (int i) const
+  {
+      if (is_varying_size())
+      {
+        throw std::runtime_error("Rank isn't fixed");
+      }
+      return dims_[i];
+  }
+
+  size_t size() const
+  {
+    if (is_varying_size())
+    {
+      throw std::runtime_error("Rank isn't fixed");
+    }
+    return dims_.size();
+  }
+
+  bool is_varying_size() const { return is_varying_size_; }
+
+  VaryingShape merge(const VaryingShape& other) const;
+
+  std::vector<int64_t> freeze() const
+  {
+    std::vector<int64_t> frozen;
+    if (is_varying_size())
+    {
+      throw std::runtime_error("Rank isn't fixed");
+    }
+    for (auto d : dims_)
+    {
+      if (!d)
+      {
+        throw std::runtime_error("Shape isn't fixed");
+      }
+      frozen.push_back(d.value());
+    }
+
+    return frozen;
+  }
+
+private:
+  bool is_varying_size_;
+  std::vector<c10::optional<int64_t>> dims_;
```

Comment:
My (slight) preference goes to `optional<vector<optional<int>>>`. It might not look beautiful, but it uses `optional` in a use case it was designed for and does not reimplement it in a custom way. IOW, in the current code a similar problem is solved in two different ways in the same place, which I don't really like that much.